src/amm.rs

```diff
@@ -28,31 +29,42 @@ macro_rules! to_u256 {
     );
 }
 
-macro_rules! to_u128 {
+macro_rules! to_balance {
     ($x:expr) => {
-        u128::try_from($x).ok()
+        Balance::try_from($x).ok()
```

Comment:
I suggest to implement this factor as follows:

Balance::try_from($x).map_err(|_| Overflow)

You can then get rid of all the unwraps and or_ok(Overflow)  where this macros is used

Reply:
Nice. Thanks.